src/messages/tests/integration.rs

```diff
@@ -0,0 +1,749 @@
+        share_commitment: share_commitment,
+    });
+    let validate_payload = Validate::validate(&payload);
+    let valid_payload = validate_payload.expect("a valid payload").clone();
+
+    let message = Message {
+        header,
+        payload: valid_payload.clone(),
+    };
+
+    let validate_message = Validate::validate(&message);
+    assert_eq!(validate_message, Err(MsgErr::SenderMustBeDealer));
+
+    // change the header
+    let header = create_valid_header(setup.dealer, setup.aggregator);
+
+    let message = Message {
+        header,
+        payload: valid_payload,
+    };
+
+    let validate_message = Validate::validate(&message);
+    assert_eq!(validate_message, Err(MsgErr::ReceiverMustBeSigner));
+
+    let participants = vec![setup.signer1];
+    let mut share_commitment = generate_share_commitment(&shares, participants);
+
+    // change the payload to have only 1 commitment
+    let payload = Payload::SharePackage(SharePackage {
+        group_public,
+        secret_share: secret_share,
+        share_commitment: share_commitment.clone(),
+    });
+    let validate_payload = Validate::validate(&payload);
+    assert_eq!(
+        validate_payload,
+        Err(MsgErr::NotEnoughCommitments(constants::MIN_SIGNERS))
+    );
+
+    // build and use too many commitments
+    for i in 2..constants::MAX_SIGNERS as u64 + 2 {
+        share_commitment.insert(
+            ParticipantId::Signer(i),
+            share_commitment.clone()[&setup.signer1],
+        );
+    }
+    let payload = Payload::SharePackage(SharePackage {
+        group_public,
+        secret_share,
+        share_commitment,
+    });
+    let validate_payload = Validate::validate(&payload);
+    assert_eq!(validate_payload, Err(MsgErr::TooManyCommitments));
+}
+
+#[test]
+fn serialize_sharepackage() {
+    let setup = basic_setup();
+
+    let (shares, _pubkeys) =
+        frost::keygen_with_dealer(setup.num_signers, setup.threshold, setup.rng.clone()).unwrap();
+
+    let header = create_valid_header(setup.dealer, setup.signer1);
+
+    let group_public = VerificationKey::from(
+        verification_key::VerificationKey::try_from(shares[0].group_public.bytes).unwrap(),
+    );
+    let secret_share = Secret(shares[0].share.value.0.to_bytes());
+
+    let participants = vec![setup.signer1];
+    let share_commitment = generate_share_commitment(&shares, participants);
+
+    let payload = Payload::SharePackage(SharePackage {
+        group_public,
+        secret_share,
+        share_commitment: share_commitment.clone(),
+    });
+
+    let message = Message {
+        header: header,
+        payload: payload.clone(),
+    };
+
+    // check general structure and header serialization/deserialization
+    serialize_message(message, setup.dealer, setup.signer1);
+
+    // check payload serialization/deserialization
+    let mut payload_serialized_bytes = bincode::serialize(&payload).unwrap();
+    // remove the padding. TODO: why is this happening ?
```

Comment:
`bincode` serializes `enums` starting with a `u32` tag indicating the enum variant. So the first four bytes are either `0`, `1`, `2`, `3` or `4`, depending on the payload type.

Not sure how to best solve this issue. Simplest way might be to change the RFC's `enum MsgType` to be `#[repr(u32)]`. To keep it `u8`, manual serialization and deserialization implementations for `Payload` might be necessary, to make sure that only one byte is serialized, and then the payload sub-type can be serialized normally.

Reply:
We can keep the RFC as `u8` by now and have `u32` in the implementation. The reason is i think we might want to change this to `u8` as it is cheaper but we can probably do it later. I made this change with a `TODO` note in https://github.com/ZcashFoundation/redjubjub/pull/114/commits/760115ed9b6f214c809aa40d8c83be443be968cf and removed all the `TODO`s from the `integration.rs` that were related to this. Also added a `msg_type` check.

Reply:
I opened a new ticket to decide what to do on this one: https://github.com/ZcashFoundation/redjubjub/issues/119